File: AOS/app/src/main/java/com/example/escapephone/core/game/ConvenienceStorePuzzleEngines.kt
package com.example.escapephone.core.game

import com.example.escapephone.core.model.ConvenienceStoreSpec
import com.example.escapephone.core.model.IncidentEvent

class ReceiptPuzzleEngine {
    var selectedItemId: String? = null; private set
    var isSolved = false; private set
    fun selectReceiptItem(itemId: String) { if (!isSolved) selectedItemId = itemId }
    fun compareReceiptPrice(itemId: String): Boolean = ConvenienceStoreSpec.receiptItems.any { it.id == itemId && it.code == ConvenienceStoreSpec.receiptAnomalyCode }
    fun submitReceiptAnswer(): Boolean {
        if (isSolved) return true
        isSolved = selectedItemId == ConvenienceStoreSpec.receiptAnomalyItemId
        return isSolved
    }
    fun completeReceiptPuzzle(): Boolean { selectedItemId = ConvenienceStoreSpec.receiptAnomalyItemId; isSolved = true; return true }
}

class BarcodePuzzleEngine {
    var selectedSegments: Map<String, String> = emptyMap(); private set
    var isSolved = false; private set
    fun selectBarcodeSegment(segment: String, value: String) { if (!isSolved) selectedSegments = selectedSegments + (segment to value) }
    fun assembleBarcodeCode(): String = ConvenienceStoreSpec.barcodeSegmentOrder.joinToString("") { selectedSegments[it].orEmpty() }
    fun submitBarcodeRule(): Boolean {
        if (isSolved) return true
        isSolved = selectedSegments["category"] == "99" && selectedSegments["shelfPosition"] == "40" && selectedSegments["arrivalOrder"] == "01"
        return isSolved
    }
    fun completeBarcodePuzzle(): Boolean { selectedSegments = mapOf("category" to "99", "shelfPosition" to "40", "arrivalOrder" to "01"); isSolved = true; return true }
}

class TiltObjectPuzzleEngine(
    var objectX: Float = 0.5f,
    var objectY: Float = 0.5f,
    private val destinationX: Float = 0.5f,
    private val destinationY: Float = 0.9f,
    private val destinationRadius: Float = 0.12f
) {
    var isSolved = false; private set
    fun applyTiltInput(deltaX: Float, deltaY: Float) { updateObjectPosition(objectX + deltaX, objectY + deltaY) }
    fun applyTouchInput(x: Float, y: Float) { updateObjectPosition(x, y) }
    fun updateObjectPosition(x: Float, y: Float) {
        objectX = x.coerceIn(0f, 1f); objectY = y.coerceIn(0f, 1f)
        val distance = kotlin.math.hypot((objectX - destinationX).toDouble(), (objectY - destinationY).toDouble())
        if (distance <= destinationRadius) isSolved = true
    }
}

class ShelfDifferencePuzzleEngine {
    var selectedDifferenceIds: List<String> = emptyList(); private set
    var isSolved = false; private set
    fun selectShelfDifference(differenceId: String) { if (!isSolved && differenceId !in selectedDifferenceIds) selectedDifferenceIds = selectedDifferenceIds + differenceId }
    fun submitShelfDifferences(): Boolean {
        if (isSolved) return true
        isSolved = ConvenienceStoreSpec.shelfDifferenceTargetIds.all { it in selectedDifferenceIds }
        return isSolved
    }
    fun completeShelfDifferencePuzzle(): Boolean { selectedDifferenceIds = ConvenienceStoreSpec.shelfDifferenceTargetIds; isSolved = true; return true }
}

class CctvPuzzleEngine {
    var recordOrder: List<String> = ConvenienceStoreSpec.cctvRecords.map { it.id }.reversed(); private set
    var playingRecordId: String? = null; private set
    var isSolved = false; private set
    fun playCctvRecord(recordId: String) { if (ConvenienceStoreSpec.cctvRecords.any { it.id == recordId }) playingRecordId = recordId }
    fun moveCctvRecordUp(index: Int) { if (!isSolved && index in recordOrder.indices && index > 0) recordOrder = recordOrder.toMutableList().also { val item = it[index]; it[index] = it[index - 1]; it[index - 1] = item } }
    fun moveCctvRecordDown(index: Int) { if (!isSolved && index in recordOrder.indices && index < recordOrder.lastIndex) recordOrder = recordOrder.toMutableList().also { val item = it[index]; it[index] = it[index + 1]; it[index + 1] = item } }
    fun submitCctvSequence(): Boolean {
        if (isSolved) return true
        val expected = ConvenienceStoreSpec.cctvRecords.sortedBy { it.correctOrder }.map { it.id }
        isSolved = recordOrder == expected
        return isSolved
    }
    fun completeCctvPuzzle(): Boolean { recordOrder = ConvenienceStoreSpec.cctvRecords.sortedBy { it.correctOrder }.map { it.id }; isSolved = true; return true }
}

class InventoryPuzzleEngine {
    var selectedItemId: String? = null; private set
    var isSolved = false; private set
    fun selectInventoryItem(itemId: String) { if (!isSolved) selectedItemId = itemId }
    fun calculateExpectedInventory(itemId: String): Int {
        val item = ConvenienceStoreSpec.inventoryItems.firstOrNull { it.id == itemId } ?: return 0
        return item.previousStock + item.arrived - item.sold
    }
    fun submitInventoryDiscrepancy(): Boolean {
        if (isSolved) return true
        isSolved = selectedItemId == ConvenienceStoreSpec.discrepancyItemId
        return isSolved
    }
    fun completeInventoryPuzzle(): Boolean { selectedItemId = ConvenienceStoreSpec.discrepancyItemId; isSolved = true; return true }
}

class CustomerPatternPuzzleEngine {
    var decodedMessage: String? = null; private set
    var isSolved = false; private set
    var selectedPurchases: List<Pair<Int, String>> = emptyList(); private set
    fun selectCustomerPurchase(cycle: Int, item: String) { if (!isSolved) selectedPurchases = selectedPurchases + (cycle to item) }
    fun decodePurchasePattern(): String = ConvenienceStoreSpec.customerPatternMessage
    fun submitCustomerPattern(candidate: String): Boolean {
        if (isSolved) return true
        isSolved = candidate == ConvenienceStoreSpec.customerPatternMessage
        if (isSolved) decodedMessage = candidate
        return isSolved
    }
    fun completeCustomerPatternPuzzle(): Boolean { decodedMessage = ConvenienceStoreSpec.customerPatternMessage; isSolved = true; return true }
}

class IncidentTimelinePuzzleEngine {
    var eventOrder: List<String> = ConvenienceStoreSpec.incidentEvents.map { it.id }.reversed(); private set
    var selectedConclusionIds: Set<String> = emptySet(); private set
    var isSolved = false; private set
    fun moveIncidentEvent(eventId: String, newIndex: Int) { if (!isSolved && eventId in eventOrder && newIndex in eventOrder.indices) eventOrder = eventOrder.toMutableList().also { val oldIndex = it.indexOf(eventId); it.removeAt(oldIndex); it.add(newIndex, eventId) } }
    var evidenceLinks: Set<Pair<String, String>> = emptySet(); private set
    fun connectEvidence(eventId: String, evidenceId: String) { if (!isSolved) evidenceLinks = evidenceLinks + (eventId to evidenceId) }
    fun selectIncidentConclusion(conclusionId: String) { if (!isSolved) selectedConclusionIds = selectedConclusionIds + conclusionId }
    fun isOrderCorrect(): Boolean = eventOrder == ConvenienceStoreSpec.incidentEvents.sortedBy { it.correctOrder }.map { it.id }
    fun areConclusionsCorrect(): Boolean = ConvenienceStoreSpec.incidentEvents.map { it.id }.all { it in selectedConclusionIds }
    fun validateIncidentTimeline(): Boolean = isOrderCorrect() && areConclusionsCorrect()
    fun completeIncidentTimelinePuzzle(): Boolean {
        eventOrder = ConvenienceStoreSpec.incidentEvents.sortedBy { it.correctOrder }.map { it.id }
        selectedConclusionIds = ConvenienceStoreSpec.incidentEvents.map { it.id }.toSet()
        isSolved = true
        return true
    }
}

File: ios/EscapePhone/Core/Game/ConvenienceStorePuzzleEngines.swift
import CoreGraphics
import Foundation

struct ReceiptPuzzleEngine {
    private(set) var selectedItemId: String?
    private(set) var isSolved = false
    mutating func selectReceiptItem(_ itemId: String) { guard !isSolved else { return }; selectedItemId = itemId }
    func compareReceiptPrice(_ itemId: String) -> Bool { ConvenienceStoreSpec.receiptItems.contains { $0.id == itemId && $0.code == ConvenienceStoreSpec.receiptAnomalyCode } }
    mutating func submitReceiptAnswer() -> Bool {
        guard !isSolved else { return true }
        isSolved = selectedItemId == ConvenienceStoreSpec.receiptAnomalyItemId
        return isSolved
    }
    mutating func completeReceiptPuzzle() -> Bool { selectedItemId = ConvenienceStoreSpec.receiptAnomalyItemId; isSolved = true; return true }
}

struct BarcodePuzzleEngine {
    private(set) var selectedSegments: [String: String] = [:]
    private(set) var isSolved = false
    mutating func selectBarcodeSegment(_ segment: String, value: String) { guard !isSolved else { return }; selectedSegments[segment] = value }
    func assembleBarcodeCode() -> String { ConvenienceStoreSpec.barcodeSegmentOrder.map { selectedSegments[$0] ?? "" }.joined() }
    mutating func submitBarcodeRule() -> Bool {
        guard !isSolved else { return true }
        isSolved = selectedSegments["category"] == "99" && selectedSegments["shelfPosition"] == "40" && selectedSegments["arrivalOrder"] == "01"
        return isSolved
    }
    mutating func completeBarcodePuzzle() -> Bool { selectedSegments = ["category": "99", "shelfPosition": "40", "arrivalOrder": "01"]; isSolved = true; return true }
}

struct TiltObjectPuzzleEngine {
    private(set) var objectX: CGFloat
    private(set) var objectY: CGFloat
    private let destinationX: CGFloat
    private let destinationY: CGFloat
    private let destinationRadius: CGFloat
    private(set) var isSolved = false

    init(objectX: CGFloat = 0.5, objectY: CGFloat = 0.5, destinationX: CGFloat = 0.5, destinationY: CGFloat = 0.9, destinationRadius: CGFloat = 0.12) {
        self.objectX = objectX; self.objectY = objectY; self.destinationX = destinationX; self.destinationY = destinationY; self.destinationRadius = destinationRadius
    }
    mutating func applyTiltInput(deltaX: CGFloat, deltaY: CGFloat) { updateObjectPosition(x: objectX + deltaX, y: objectY + deltaY) }
    mutating func applyTouchInput(x: CGFloat, y: CGFloat) { updateObjectPosition(x: x, y: y) }
    mutating func updateObjectPosition(x: CGFloat, y: CGFloat) {
        objectX = min(max(x, 0), 1); objectY = min(max(y, 0), 1)
        let distance = hypot(objectX - destinationX, objectY - destinationY)
        if distance <= destinationRadius { isSolved = true }
    }
}

struct ShelfDifferencePuzzleEngine {
    private(set) var selectedDifferenceIds: [String] = []
    private(set) var isSolved = false
    mutating func selectShelfDifference(_ differenceId: String) { guard !isSolved, !selectedDifferenceIds.contains(differenceId) else { return }; selectedDifferenceIds.append(differenceId) }
    mutating func submitShelfDifferences() -> Bool {
        guard !isSolved else { return true }
        isSolved = ConvenienceStoreSpec.shelfDifferenceTargetIds.allSatisfy { selectedDifferenceIds.contains($0) }
        return isSolved
    }
    mutating func completeShelfDifferencePuzzle() -> Bool { selectedDifferenceIds = ConvenienceStoreSpec.shelfDifferenceTargetIds; isSolved = true; return true }
}

struct CctvPuzzleEngine {
    private(set) var recordOrder: [String] = Array(ConvenienceStoreSpec.cctvRecords.map(\.id).reversed())
    private(set) var playingRecordId: String?
    private(set) var isSolved = false
    mutating func playCctvRecord(_ recordId: String) { if ConvenienceStoreSpec.cctvRecords.contains(where: { $0.id == recordId }) { playingRecordId = recordId } }
    mutating func moveCctvRecordUp(_ index: Int) { guard !isSolved, recordOrder.indices.contains(index), index > 0 else { return }; recordOrder.swapAt(index, index - 1) }
    mutating func moveCctvRecordDown(_ index: Int) { guard !isSolved, recordOrder.indices.contains(index), index < recordOrder.count - 1 else { return }; recordOrder.swapAt(index, index + 1) }
    mutating func submitCctvSequence() -> Bool {
        guard !isSolved else { return true }
        let expected = ConvenienceStoreSpec.cctvRecords.sorted { $0.correctOrder < $1.correctOrder }.map(\.id)
        isSolved = recordOrder == expected
        return isSolved
    }
    mutating func completeCctvPuzzle() -> Bool { recordOrder = ConvenienceStoreSpec.cctvRecords.sorted { $0.correctOrder < $1.correctOrder }.map(\.id); isSolved = true; return true }
}

struct InventoryPuzzleEngine {
    private(set) var selectedItemId: String?
    private(set) var isSolved = false
    mutating func selectInventoryItem(_ itemId: String) { guard !isSolved else { return }; selectedItemId = itemId }
    func calculateExpectedInventory(_ itemId: String) -> Int {
        guard let item = ConvenienceStoreSpec.inventoryItems.first(where: { $0.id == itemId }) else { return 0 }
        return item.previousStock + item.arrived - item.sold
    }
    mutating func submitInventoryDiscrepancy() -> Bool {
        guard !isSolved else { return true }
        isSolved = selectedItemId == ConvenienceStoreSpec.discrepancyItemId
        return isSolved
    }
    mutating func completeInventoryPuzzle() -> Bool { selectedItemId = ConvenienceStoreSpec.discrepancyItemId; isSolved = true; return true }
}

struct CustomerPatternPuzzleEngine {
    private(set) var selectedPurchases: [(cycle: Int, item: String)] = []
    private(set) var decodedMessage: String?
    private(set) var isSolved = false
    mutating func selectCustomerPurchase(_ cycle: Int, item: String) { guard !isSolved else { return }; selectedPurchases.append((cycle, item)) }
    func decodePurchasePattern() -> String { ConvenienceStoreSpec.customerPatternMessage }
    mutating func submitCustomerPattern(_ candidate: String) -> Bool {
        guard !isSolved else { return true }
        isSolved = candidate == ConvenienceStoreSpec.customerPatternMessage
        if isSolved { decodedMessage = candidate }
        return isSolved
    }
    mutating func completeCustomerPatternPuzzle() -> Bool { decodedMessage = ConvenienceStoreSpec.customerPatternMessage; isSolved = true; return true }
}

struct IncidentTimelinePuzzleEngine {
    private(set) var eventOrder: [String] = Array(ConvenienceStoreSpec.incidentEvents.map(\.id).reversed())
    private(set) var selectedConclusionIds: Set<String> = []
    private(set) var evidenceLinks: Set<String> = []
    private(set) var isSolved = false
    mutating func moveIncidentEvent(_ eventId: String, newIndex: Int) { guard !isSolved, let oldIndex = eventOrder.firstIndex(of: eventId), eventOrder.indices.contains(newIndex) else { return }; let value = eventOrder.remove(at: oldIndex); eventOrder.insert(value, at: newIndex) }
    mutating func connectEvidence(_ eventId: String, evidenceId: String) { guard !isSolved else { return }; evidenceLinks.insert("\(eventId)->\(evidenceId)") }
    mutating func selectIncidentConclusion(_ conclusionId: String) { guard !isSolved else { return }; selectedConclusionIds.insert(conclusionId) }
    func isOrderCorrect() -> Bool { eventOrder == ConvenienceStoreSpec.incidentEvents.sorted { $0.correctOrder < $1.correctOrder }.map(\.id) }
    func areConclusionsCorrect() -> Bool { ConvenienceStoreSpec.incidentEvents.map(\.id).allSatisfy { selectedConclusionIds.contains($0) } }
    func validateIncidentTimeline() -> Bool { isOrderCorrect() && areConclusionsCorrect() }
    mutating func completeIncidentTimelinePuzzle() -> Bool {
        eventOrder = ConvenienceStoreSpec.incidentEvents.sorted { $0.correctOrder < $1.correctOrder }.map(\.id)
        selectedConclusionIds = Set(ConvenienceStoreSpec.incidentEvents.map(\.id))
        isSolved = true
        return true
    }
}
